File: JavaScript/DeleteEmptyComponents.js
//Author-Autodesk Inc.
//Description-Delete empty components from an assembly.
/*globals adsk*/
(function () {

    'use strict';

    if (adsk.debug === true) {
        /*jslint debug: true*/
        debugger;
        /*jslint debug: false*/
    }
    
    var ui;
    try {
        
        var app = adsk.core.Application.get();
        ui = app.userInterface;
        
        // Get all components in the active design.
        var product = app.activeProduct;
        var design = adsk.fusion.Design(product);
        var title = 'Delete Empty Components';
        if (!design) {
            ui.messageBox('No active Fusion design', title);
            adsk.terminate();
            return;
        }
        var components = design.allComponents;

        // Find all of the empty components.
        // It is empty if it has no occurrences, bodies, featres, sketches, or construction.
        var root = design.rootComponent;
        var nComponents = components.count;
        var componentsToDelete = [];
        var i, component;
        for (i = 0; i < nComponents; ++i) {
            component = components.item(i);

            // Skip the root component.
            if (root.equals(component)) {
                continue;
            }

            if (component.occurrences.count === 0
                    && component.bRepBodies.count === 0
                    && component.features.count === 0
                    && component.sketches.count === 0
                    && component.constructionPlanes.count === 0
                    && component.constructionAxes.count === 0
                    && component.constructionPoints.count === 0) {
                componentsToDelete.push(component);
            }
        }

        // Delete all immediate occurrences of the empty components.
        var deletedComponents = [], name, occurrences, j, occurrence, uniqueOccurrences, k;
        for (i = 0; i < componentsToDelete.length; ++i) {
            component = componentsToDelete[i];

            // Get the name first because deleting the final Occurrence will delete the Component.
            name = component.name;

            // Build a list of unique immediate occurrences of the component.
            occurrences = root.allOccurrencesByComponent(component);
            uniqueOccurrences = [];
            for (j = 0; j < occurrences.count; ++j) {
                occurrence = occurrences.item(j).nativeObject;
                for (k = 0; k < uniqueOccurrences.length; ++k) {
                    if (occurrence.equals(uniqueOccurrences[k])) {
                        break;
                    }
                }
                if (k === uniqueOccurrences.length) {
                    uniqueOccurrences.push(occurrence);
                }
            }

            // Delete them.
            for (j = 0; j < uniqueOccurrences.length; ++j) {
                uniqueOccurrences[j].deleteMe();
            }
            deletedComponents.push(name);
        }

        var msg;
        if (deletedComponents.length === 0) {
            msg = 'No empty components found.';
        } else {
            msg = deletedComponents.length + ' empty component' + (deletedComponents.length > 1 ? 's' : '') + ' deleted';
            msg += '\n\n';
            for (i = 0; i < deletedComponents.length; ++i) {
                msg += '\n' + deletedComponents[i];
            }
        }
        ui.messageBox(msg, title);

    } 
    catch (e) {
        if (ui) {
            ui.messageBox('Failed : ' + (e.description ? e.description : e));
        }
    }

    adsk.terminate();
}());

File: Python/DeleteEmptyComponents.py
#Author-Autodesk Inc.
#Description-Delete empty components from an assembly.

import adsk, adsk.core, adsk.fusion, traceback

def main():
    ui = None
    try:
        app = adsk.core.Application.get()
        ui = app.userInterface
        
        # Get all components in the active design.
        product = app.activeProduct
        design = product
        title = 'Delete Empty Components'
        if not design:
            ui.messageBox('No active Fusion design', title)
            return

        components = design.allComponents

        # Find all of the empty components.
        # It is empty if it has no occurrences, bodies, featres, sketches, or construction.
        root = design.rootComponent
        componentsToDelete = []

        for component in components:

            # Skip the root component.
            if root == component:
                continue

            if len(component.occurrences) == 0 \
                and len(component.bRepBodies) == 0 \
                and len(component.features) == 0 \
                and len(component.sketches) == 0 \
                and len(component.constructionPlanes) == 0 \
                and len(component.constructionAxes) == 0 \
                and len(component.constructionPoints) == 0:

                componentsToDelete.append(component)

        # Delete all immediate occurrences of the empty components.
        deletedComponents = []
        global k
        k = 0
        for component in componentsToDelete:

            # Get the name first because deleting the final Occurrence will delete the Component.
            name = component.name

            # Build a list of unique immediate occurrences of the component.
            occurrences = root.allOccurrencesByComponent(component)
            uniqueOccurrences = []
            for occurrence in occurrences:
                
                for k in range(0, len(uniqueOccurrences)):
                    if occurrence is uniqueOccurrences[k]:
                        break
                if k == len(uniqueOccurrences):
                    uniqueOccurrences.append(occurrence)

            # Delete them.
            for uniqueOccurrencesI in uniqueOccurrences:
                uniqueOccurrencesI.deleteMe()

            deletedComponents.append(name)

        if len(deletedComponents) == 0:
            msg = 'No empty components found.'
        else:
            if len(deletedComponents) > 1:
                msg = str(len(deletedComponents)) + ' empty component' + 's'
            else:
                msg = str(len(deletedComponents)) + ' empty component' + ' deleted'
            msg += '\n\n'
            for deletedComponentI in deletedComponents:
                msg += '\n' + deletedComponentI

        ui.messageBox(msg, title)

    except:
        if ui:
            ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

main()
